### PR TITLE
Added CinemaHall, MovieSession, Order and User controllers

### PR DESCRIPTION
**Pull request by mikhail771**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dto/OrderRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package com.dev.cinema.dto;
+
+import java.util.List;
+
+public class OrderRequestDto {
```
Comment (OleksiiTromsa) [inline src/main/java/com/dev/cinema/dto/OrderRequestDto.java]: You don't use this class in your app

File: src/main/java/com/dev/cinema/dto/UserResponseDto.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package com.dev.cinema.dto;
+
+public class UserResponseDto {
+    private Long id;
+    private String email;
+    private String password;
```
Comment (OleksiiTromsa) [inline src/main/java/com/dev/cinema/dto/UserResponseDto.java]: I'm not sure, but maybe it is a bad idea to show password to anybody?

**Review by OleksiiTromsa — COMMENTED** [review]
Good job

File: src/main/java/com/dev/cinema/dto/OrderRequestDto.java (hunk 1 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dto/OrderRequestDto.java]: done

File: src/main/java/com/dev/cinema/dto/UserResponseDto.java (hunk 2 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dto/UserResponseDto.java]: done

File: src/main/java/com/dev/cinema/controller/CinemaHallController.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private CinemaHallMapper cinemaHallMapper;
+
+    public CinemaHallController(CinemaHallService cinemaHallService,
+                                CinemaHallMapper cinemaHallMapper) {
+        this.cinemaHallService = cinemaHallService;
+        this.cinemaHallMapper = cinemaHallMapper;
+    }
+
+    @PostMapping
+    public CinemaHall add(@RequestBody CinemaHallRequestDto dto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/CinemaHallController.java]: let's not return the entity itself (can be applied to all the methods where you do so)

return void or at least ResponseDto

File: src/main/java/com/dev/cinema/controller/MovieSessionController.java [hunk 4]
```diff
@@ -16,7 +16,7 @@
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
-@RequestMapping("/movie-sessions")
+@RequestMapping("/movie-session")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/MovieSessionController.java]: why have you changed this?

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    @PostMapping("/complete")
+    public Order completeOrder(@RequestParam String email) {
+        User user = userService.findByEmail(email).get();
+        ShoppingCart cart = shoppingCartService.getByUser(user);
+        return orderService.completeOrder(cart.getTickets(), user);
+    }
+
+    @GetMapping
+    public List<OrderResponseDto> getOrderHistory(@RequestParam Long userId) {
+        return orderService.getOrderHistory(userService.findById(userId).get()).stream()
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/OrderController.java]:
```suggestion
        return orderService.getOrderHistory(userService.findById(userId)).stream()
```
let's make this method return User, not Optional

File: src/main/java/com/dev/cinema/controller/CinemaHallController.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/cinema-halls")
+public class CinemaHallController {
+    private CinemaHallService cinemaHallService;
+    private CinemaHallMapper cinemaHallMapper;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/CinemaHallController.java]: add `final` keyword

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private ShoppingCartService shoppingCartService;
+    private UserService userService;
+    private OrderService orderService;
+    private OrderMapper orderMapper;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/OrderController.java]: same

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 8]
```diff
@@ -56,4 +57,14 @@ public MovieSession add(MovieSession movieSession) {
             return query.getResultList();
         }
     }
+
+    @Override
+    public Optional<MovieSession> findById(Long id) {
+        try (Session session = sessionFactory.openSession()) {
+            return Optional.of(session.get(MovieSession.class, id));
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: what happens if we call `Optional.of(null)`?

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 9]
```diff
@@ -49,4 +49,14 @@ public User add(User user) {
...
         }
     }
+
+    @Override
+    public Optional<User> findById(Long id) {
+        try (Session session = sessionFactory.openSession()) {
+            Query<User> query = session.createQuery("from User "
+                    + "where id = :id ", User.class);
+            query.setParameter("id", id);
+            return query.uniqueResultOptional();
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: let's use `session.get()`

File: src/main/java/com/dev/cinema/service/mapper/CinemaHallMapper.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.dev.cinema.dto.CinemaHallRequestDto;
+import com.dev.cinema.dto.CinemaHallResponseDto;
+import com.dev.cinema.model.CinemaHall;
+import com.dev.cinema.service.CinemaHallService;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CinemaHallMapper {
+    private final CinemaHallService cinemaHallService;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/CinemaHallMapper.java]: do you use this dependency?

File: src/main/java/com/dev/cinema/service/mapper/OrderMapper.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import com.dev.cinema.service.UserService;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Component;
+
+@Component
+public class OrderMapper {
+    private final TicketMapper ticketMapper;
+    private final UserService userService;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/OrderMapper.java]: same

File: src/main/java/com/dev/cinema/service/mapper/ShoppingCartMapper.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+@Component
+public class ShoppingCartMapper {
+    private final TicketMapper ticketMapper;
+
+    public ShoppingCartMapper(TicketMapper ticketMapper) {
+        this.ticketMapper = ticketMapper;
+    }
+
+    public ShoppingCartResponseDto mapToDto(ShoppingCart cart) {
+        ShoppingCartResponseDto dto = new ShoppingCartResponseDto(cart.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/ShoppingCartMapper.java]: let's set id with setter as well?

File: src/main/java/com/dev/cinema/dto/ShoppingCartResponseDto.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+public class ShoppingCartResponseDto {
+    private Long userId;
+    private List<TicketResponseDto> tickets;
+
+    public ShoppingCartResponseDto() {
+    }
+
+    public ShoppingCartResponseDto(Long userId) {
+        this.userId = userId;
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dto/ShoppingCartResponseDto.java]: and remove these constructors

File: src/main/java/com/dev/cinema/service/mapper/TicketMapper.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+import com.dev.cinema.model.Ticket;
+import com.dev.cinema.service.MovieSessionService;
+import com.dev.cinema.service.UserService;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TicketMapper {
+    private final MovieSessionService movieSessionService;
+    private final MovieSessionMapper movieSessionMapper;
+    private final UserService userService;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/mapper/TicketMapper.java]: do we need all these dependencies?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/controller/CinemaHallController.java (hunk 3 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/controller/CinemaHallController.java]: done

File: src/main/java/com/dev/cinema/controller/OrderController.java (hunk 5 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/controller/OrderController.java]: done

File: src/main/java/com/dev/cinema/controller/CinemaHallController.java (hunk 6 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/controller/CinemaHallController.java]: done

File: src/main/java/com/dev/cinema/controller/OrderController.java (hunk 7 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/controller/OrderController.java]: done

File: src/main/java/com/dev/cinema/service/mapper/CinemaHallMapper.java (hunk 10 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/service/mapper/CinemaHallMapper.java]: done

File: src/main/java/com/dev/cinema/service/mapper/OrderMapper.java (hunk 11 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/service/mapper/OrderMapper.java]: done

File: src/main/java/com/dev/cinema/service/mapper/ShoppingCartMapper.java (hunk 12 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/service/mapper/ShoppingCartMapper.java]: done

File: src/main/java/com/dev/cinema/service/mapper/TicketMapper.java (hunk 14 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/service/mapper/TicketMapper.java]: done

File: src/main/java/com/dev/cinema/dto/ShoppingCartResponseDto.java (hunk 13 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dto/ShoppingCartResponseDto.java]: done

File: src/main/java/com/dev/cinema/controller/MovieSessionController.java (hunk 4 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/controller/MovieSessionController.java]: done

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+                           OrderService orderService,
+                           OrderMapper orderMapper) {
+        this.shoppingCartService = shoppingCartService;
+        this.userService = userService;
+        this.orderService = orderService;
+        this.orderMapper = orderMapper;
+    }
+
+    @PostMapping("/complete")
+    public Order completeOrder(@RequestParam String email) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/OrderController.java]:
```suggestion
    public void completeOrder(@RequestParam String email) {
```

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 8 above)
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: again, what happens if we call Optional.of(null)?

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java (hunk 9 above)
Comment (KseniiaMakarova) [reply src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: the previous comment is not fixed

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/controller/OrderController.java (hunk 15 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/controller/OrderController.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 8 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: Will be NPE. Sorry, i'm blind

Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: Sorry

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java (hunk 9 above)
Comment (mikhail771) [reply src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: Sorry. Done

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**